actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb

```diff
@@ -0,0 +1,87 @@
+require 'active_support/core_ext/object/try'
+
+module ActionView
+  module CollectionCaching # :nodoc:
+    extend ActiveSupport::Concern
+
+    included do
+      # Fallback cache store if Action View is used without Rails.
+      # Otherwise overriden in Railtie to use Rails.cache.
+      mattr_accessor(:collection_cache) { ActiveSupport::Cache::MemoryStore.new }
+    end
+
+    private
+
+    def cache_collection_render
+      return yield unless cache_collection?
+
+      keyed_collection = collection_by_cache_keys
+      partial_cache = collection_cache.read_multi(*keyed_collection.keys)
+
+      unless partial_cache.empty?
+        ActionView::Base.logger.try :debug, "#{partial_cache.size} partials " \
+          "multi read from cache."
+      end
+
+      @collection = keyed_collection.reject { |key, _| partial_cache.key?(key) }.values
+      rendered_partials = @collection.any? ? yield.dup : []
+
+      unless rendered_partials.empty?
+        ActionView::Base.logger.try :debug, "#{rendered_partials.size} partials" \
+          " will be written to cache."
+      end
+
+      fetch_or_cache_partial(partial_cache, order_by: keyed_collection.each_key) do
+        rendered_partials.shift
+      end
+    end
+
+    def cache_collection?
+      @options.fetch(:cache, automatic_cache_eligible?)
+    end
+
+    def automatic_cache_eligible?
+      single_template_render? && !callable_cache_key? && template_eligible?
+    end
+
+    def single_template_render?
+      @template # Template is only set when a collection renders one template.
+    end
+
+    def callable_cache_key?
+      @options[:cache].respond_to?(:call)
+    end
+
+    def template_eligible?
+      @template.eligible_for_collection_caching?(as: @options[:as]).tap do
+        ActionView::Base.logger.try :debug,
+          "Collection automatically caches #{@template.virtual_path}, so next render uses a multi read for any templates rendered now."
```

Comment:
Okay, I've been looking more into logging and captured this from the logging test:

```
Collection automatically caches customers/_customer, so next render uses a multi read for any templates rendered now.
Cache digest for /Users/kasperhansen/Documents/code/rails/actionpack/test/fixtures/customers/_customer.html.erb: 4184ab71db6849621a4d8820fcd2c0ad
Cache digest for /Users/kasperhansen/Documents/code/rails/actionpack/test/fixtures/collection_cache/index.html.erb: 8f52a1ba24922b018905574fbb082059
Cache digest for /Users/kasperhansen/Documents/code/rails/actionpack/test/fixtures/customers/_customer.html.erb: 4184ab71db6849621a4d8820fcd2c0ad
Cache read: views/david/1/4184ab71db6849621a4d8820fcd2c0ad
Cache write: views/david/1/4184ab71db6849621a4d8820fcd2c0ad
1 partials will be written to cache.
Cache write: views/david/1/8f52a1ba24922b018905574fbb082059

Collection automatically caches customers/_customer, so next render uses a multi read for any templates rendered now.
1 partials multi read from cache.
Cache read: views/david/2/4184ab71db6849621a4d8820fcd2c0ad
Cache write: views/david/2/4184ab71db6849621a4d8820fcd2c0ad
Cache read: views/david/3/4184ab71db6849621a4d8820fcd2c0ad
Cache write: views/david/3/4184ab71db6849621a4d8820fcd2c0ad
2 partials will be written to cache.
Cache write: views/david/2/8f52a1ba24922b018905574fbb082059
Cache write: views/david/3/8f52a1ba24922b018905574fbb082059

Collection automatically caches customers/_customer, so next render uses a multi read for any templates rendered now.
3 partials multi read from cache.
```

It seems the cache store doesn't log when using multi read. But the "3 partials multi read from cache." lacks information to be really useful, I think.